tests/filesystem.h

```diff
@@ -0,0 +1,13 @@
+#if !defined(__has_include)
+// Assume experimental if compiler doesn't have __has_include
+#include <experimental/filesystem>
+namespace fs = std::experimental::filesystem;
+#elif __has_include(<filesystem>)
+#include <filesystem>
+namespace fs = std::filesystem;
```

Comment:
I might be underthinking it, but: why do we need to discover `std::experimental::filesystem` at all? C++17 conformant compilers should *always* have `std::filesystem`, so shouldn't we be able to skip this + the CMake hacks? Or are there C++17 compilers that don't have `std::filesystem` support?

Reply:
GCC 7 uses `std::experimental::filesystem` and also requires a specific link to `libstdc++fs`, which I found out after running it through CI without all this. See this question also https://stackoverflow.com/questions/48729328/link-errors-using-filesystem-members-in-c17

So while C++17 gives us nicer filesystem support, it can sometimes be a pain to get everything compiled and linked correctly 🙃

Reply:
Ugh. Apparently GCC 8 promotes it to `std::filesystem`, which was last year's release. That's probably a little too new to require. 

Let's stick with this header for now, although I wouldn't mind if it was renamed to `filesystem_compat.h` or something similar to indicate that we're hacking around misbehaving standards support in different compilers. A comment at the top would also be good 🙂